source/common/router/rds_impl.cc

```diff
@@ -154,21 +155,24 @@ RouteConfigProviderManagerImpl::routeConfigProviders() {
 };
 
 Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteConfigProvider(
-    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const envoy::api::v2::filter::Rds& rds, Upstream::ClusterManager& cm, Stats::Scope& scope,
     const std::string& stat_prefix, Init::Manager& init_manager) {
 
-  // RdsRouteConfigProviders are unique based on their <route_config_name>MAP_CONCATENATOR<cluster>.
+  // RdsRouteConfigProviders are unique based on their
+  // <route_config_name>MAP_CONCATENATOR<serialized RDS config>.
+  // TODO(htuch): Full serialization here gives large IDs, could get away with a
+  // strong hash instead.
   const std::string manager_identifier =
-      config.getString("route_config_name") + MAP_CONCATENATOR + config.getString("cluster");
+      rds.route_config_name() + MAP_CONCATENATOR + rds.SerializeAsString();
```

Comment:
Yep, fine with `unordered_map` here, just wanted to avoid an additional hash. I've opened https://github.com/lyft/envoy/issues/1452 to track this for the general xDS API update checks.